### PR TITLE
Support Bitbucket

### PR DESCRIPTION
**Pull request by zot**
This PR adds support for Bitbucket, as well as some source documentation and refactoring needed for Bitbucket support -- Bitbucket diverges from many of the similarities between GitHub and GitLab.

### CODE REVIEW

**Comment by zot** [comment]
Hmm, tests are failing but not locally -- I'll get that working...

File: src/webui/WebUI.jl [hunk 1]
```diff
@@ -92,6 +115,11 @@ include("routes/callback.jl")
 include("routes/select.jl")
 include("routes/register.jl")
 include("routes/status.jl")
+include("routes/bitbucket.jl")
+
+auth_href(::Provider{GitLabAPI}) = "$(ROUTES[:AUTH])?provider=gitlab"
+auth_href(::Provider{GitHubAPI}) = "$(ROUTES[:AUTH])?provider=github"
+auth_href(::Provider{BitbucketAPI}) = "$(ROUTES[:BITBUCKET])/auth"
```
Comment (nkottary) [inline src/webui/WebUI.jl:122]: Can't we use the same format `/auth/?provider=bitbucket`?

File: src/webui/WebUI.jl [hunk 2]
```diff
@@ -102,16 +130,22 @@ function init_registry()
...
     owner, name = splitrepo(url)
-    repo = @gf get_repo(forge, owner, name)
-    repo === nothing && error("Registry lookup failed")
-
+    repo = nothing
+    try
+    	# not using @gf here so we can log the error
+        repo, _ = get_repo(forge, owner, name)
+    catch err
+        error("Registry lookup failed$(err === nothing ? "" : err)")
```
Comment (nkottary) [inline src/webui/WebUI.jl:147]: We don't need the `nothing` check here. `err` will never be `nothing`.

File: src/webui/WebUI.jl (hunk 1 above)
Comment (zot) [reply src/webui/WebUI.jl:122]: No, Bitbucket puts its own query parameters in the URL -- it was probably a bad idea in the first place to assume it was safe to use query parameters

File: src/webui/WebUI.jl (hunk 2 above)
Comment (zot) [reply src/webui/WebUI.jl]: fixed in the latest commit

File: src/webui/WebUI.jl (hunk 1 above)
Comment (zot) [reply src/webui/WebUI.jl:122]: I did try that in the beginning, btw -- if I remember correctly, bitbucket just replaced our parameters with its own

File: src/webui/gitutils.jl [hunk 3]
```diff
@@ -40,130 +66,180 @@ is_success(res::AuthFailure) = false
...
+    ((@gf_bool @mock is_collaborator(u.forge, repo.owner.login, repo.name, u.user.login)) ||
+        (@gf_bool @mock is_collaborator(jforge, repo.owner.login, repo.name, u.user.login))) &&
+        return AuthSuccess()
+    # If the repo is not in an org, the user does not have sufficient access to create a PR
+    repo.organization === nothing &&
+        return AuthFailure("User $(u.user.login) is not a collaborator on repo $(repo.name)")
+    # Otherwise the user must be a member of the organization
+    # verify with user's connection, falling back to JuliaHub's connection
+    (something((@gf_q @mock is_member(u.forge, repo.organization.login, u.user.login)), false) ||
+        something((@gf_q @mock is_member(jforge, repo.organization.login, u.user.login)), false)) &&
```
Comment (nkottary) [inline src/webui/gitutils.jl:92]: These two lines can be simplified. `something((@gf_q f()), false)` is same as `@gf_bool f()`.

File: src/webui/gitutils.jl [hunk 4]
```diff
@@ -40,130 +66,180 @@ is_success(res::AuthFailure) = false
...
-    forge = PROVIDERS["gitlab"].client
-    result = nothing
+    forge = provider(repo).client
+    lasterr = nothing
+    for file in Base.project_names
+        try
+            fc, _ = get_file_contents(forge, repo.id, joinpath(subdir, file); ref=ref)
+            return decodeb64(fc.content)
+        catch err
+            lasterr = err
```
Comment (nkottary) [inline src/webui/gitutils.jl:172]:
```suggestion
            lasterr = (err, catch_backtrace())
```

File: src/webui/gitutils.jl (hunk 3 above)
Comment (zot) [reply src/webui/gitutils.jl]: pushed that change

**Review by nkottary — APPROVED** [review]

**Pull request merged**